File: scripts/observer_scan.py
#!/usr/bin/env python3
# DEPRECATED: Python implementation is preserved for compatibility.
# Use scripts/observer_scan.ts for active execution.
import argparse
import json
import re
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Dict, List, Optional

from equal_rank_nudge_bot import evaluate_state_file

ROOT = Path(__file__).resolve().parents[1]
PIPELINES = ROOT / ".state" / "pipelines"
WAITING = {"planned", "proposal_pending", "waiting_reply"}
EXCLUDED_TOKENS = {"demo", "test", "sample", "데모", "테스트", "샘플"}


def parse_iso(value: Optional[str]) -> Optional[datetime]:
    if not value:
        return None
    try:
        return datetime.fromisoformat(value.replace("Z", "+00:00"))
    except Exception:
        return None


def age_minutes(ts: Optional[datetime]) -> Optional[int]:
    if ts is None:
        return None
    now = datetime.now(timezone.utc)
    delta = now - ts.astimezone(timezone.utc)
    return max(0, int(delta.total_seconds() // 60))


def infer_minutes_since_reply(payload: Dict[str, Any], state_path: Path) -> int:
    nudge = payload.get("nudge", {})
    candidates = [
        parse_iso(nudge.get("last_user_reply_at")),
        parse_iso(payload.get("updated_at")),
        parse_iso(payload.get("created_at")),
    ]
    for c in candidates:
        mins = age_minutes(c)
        if mins is not None:
            return mins
    stat = state_path.stat()
    fallback = datetime.fromtimestamp(stat.st_mtime, tz=timezone.utc)
    return age_minutes(fallback) or 0


def parse_bool(value: Any, default: bool = False) -> bool:
    if isinstance(value, bool):
        return value
    if value is None:
        return default
    if isinstance(value, (int, float)):
        return bool(value)
    if isinstance(value, str):
        normalized = value.strip().lower()
        if normalized in {"1", "true", "yes", "y", "on"}:
            return True
        if normalized in {"0", "false", "no", "n", "off", ""}:
            return False
    return default


def parse_int(value: Any, default: int = 0) -> int:
    try:
        return int(value)
    except (TypeError, ValueError):
        return default


def tokenize(value: Any) -> List[str]:
    if value is None:
        return []
    text = str(value).lower()
    return [token for token in re.split(r"[^0-9a-z가-힣]+", text) if token]


def is_demo_test_sample(payload: Dict[str, Any], state_path: Path) -> bool:
    candidates = [
        payload.get("job_id"),
        state_path.stem,
        payload.get("spec_path"),
        payload.get("spawn_path"),
        payload.get("spawn_ready_path"),
        payload.get("artifacts"),
    ]
    for candidate in candidates:
        tokens = tokenize(candidate)
        if any(token in EXCLUDED_TOKENS for token in tokens):
            return True
    return False


def build_skip_result(
    payload: Dict[str, Any],
    state_file: Path,
    status: str,
    minutes_since_reply: int,
    priority: int,
    track: bool,
    reason: str,
) -> Dict[str, Any]:
    return {
        "job_id": payload.get("job_id"),
        "state_file": str(state_file),
        "should_nudge": False,
        "reason": reason,
        "message": None,
        "minutes_since_reply": minutes_since_reply,
        "status": status,
        "priority": priority,
        "observer_track": track,
    }


def scan(apply: bool, seed: Optional[int]) -> Dict[str, Any]:
    PIPELINES.mkdir(parents=True, exist_ok=True)
    results: List[Dict[str, Any]] = []

    for state_file in sorted(PIPELINES.glob("*.json")):
        payload = json.loads(state_file.read_text(encoding="utf-8"))
        status = payload.get("status", "planned")
        if status not in WAITING:
            continue

        observer = payload.get("observer", {})
        track = parse_bool(observer.get("track"), default=False)
        priority = parse_int(observer.get("priority"), default=0)
        minutes = infer_minutes_since_reply(payload, state_file)

        if is_demo_test_sample(payload, state_file):
            results.append(
                build_skip_result(
                    payload=payload,
                    state_file=state_file,
                    status=status,
                    minutes_since_reply=minutes,
                    priority=priority,
                    track=track,
                    reason="데모/테스트/샘플 제외",
                )
            )
            continue

        if not track:
            results.append(
                build_skip_result(
                    payload=payload,
                    state_file=state_file,
                    status=status,
                    minutes_since_reply=minutes,
                    priority=priority,
                    track=track,
                    reason="observer.track 꺼짐",
                )
            )
            continue

        result = evaluate_state_file(
            state_file=state_file,
            minutes_since_reply=minutes,
            seed=seed,
            apply=apply,
        )
        result["minutes_since_reply"] = minutes
        result["status"] = status
        result["priority"] = priority
        result["observer_track"] = track
        results.append(result)

    eligible = [r for r in results if r.get("should_nudge")]
    eligible.sort(
        key=lambda x: (
            -parse_int(x.get("priority"), default=0),
            -parse_int(x.get("minutes_since_reply"), default=0),
            str(x.get("job_id") or ""),
        )
    )

    return {
        "checked": len(results),
        "eligible": len(eligible),
        "top": eligible[0] if eligible else None,
        "results": results,
    }


def main() -> None:
    p = argparse.ArgumentParser(description="오 분 관찰자 스캔 러너")
    p.add_argument("--apply", action="store_true")
    p.add_argument("--seed", type=int)
    args = p.parse_args()

    print(json.dumps(scan(apply=args.apply, seed=args.seed), ensure_ascii=False))


if __name__ == "__main__":
    main()

File: scripts/observer_scan.ts
import { existsSync, mkdirSync, readdirSync, readFileSync, statSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { evaluateStateFile } from "./equal_rank_nudge_bot";
import {
  buildActionPlanFromPayload,
  normalizeObserverHints,
  parseActionKind,
  parseNumber,
} from "./_observer_tooling";

type StatePayload = Record<string, unknown>;

type ScanResult = Record<string, unknown>;

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(__dirname, "..");
const PIPELINES = path.join(ROOT, ".state", "pipelines");
const WAITING = new Set(["planned", "proposal_pending", "waiting_reply"]);
const EXCLUDED_TOKENS = new Set(["demo", "test", "sample", "데모", "테스트", "샘플"]);

function parseIso(value: unknown): Date | null {
  if (typeof value !== "string") return null;
  try {
    const date = new Date(value.replace("Z", "+00:00"));
    if (Number.isNaN(date.getTime())) return null;
    return date;
  } catch {
    return null;
  }
}

function ageMinutes(ts: Date | null): number | null {
  if (!ts) return null;
  const now = new Date();
  return Math.max(0, Math.floor((now.getTime() - ts.getTime()) / 60000));
}

function inferMinutesSinceReply(payload: StatePayload, statePath: string): number {
  const nudge = (payload.nudge as Record<string, unknown>) ?? {};
  const candidates: Array<unknown> = [
    nudge.last_user_reply_at,
    payload.updated_at,
    payload.created_at,
  ];
  for (const candidate of candidates) {
    const parsed = parseIso(candidate);
    const mins = ageMinutes(parsed);
    if (mins !== null) {
      return mins;
    }
  }

  const fallback = statSync(statePath).mtime;
  return ageMinutes(fallback) ?? 0;
}

function tokenize(value: unknown): string[] {
  if (value == null) return [];
  const text = String(value).toLowerCase();
  return text.split(/[^0-9a-z가-힣]+/g).filter(Boolean);
}

function isDemoTestSample(payload: StatePayload, statePath: string): boolean {
  const candidates = [
    payload.job_id,
    path.basename(statePath, ".json"),
    payload.spec_path,
    payload.spawn_path,
    payload.spawn_ready_path,
    payload.artifacts,
  ];
  for (const candidate of candidates) {
    const tokens = tokenize(candidate);
    if (tokens.some((token) => EXCLUDED_TOKENS.has(token))) return true;
  }
  return false;
}

function buildBaseResult(
  payload: StatePayload,
  stateFile: string,
  status: string,
  minutesSinceReply: number,
  observer: ReturnType<typeof normalizeObserverHints>,
): ScanResult {
  return {
    job_id: payload.job_id,
    state_file: stateFile,
    minutes_since_reply: minutesSinceReply,
    status,
    priority: observer.priority,
    observer_track: observer.track,
    action_plan: buildActionPlanFromPayload(payload, ROOT),
  };
}

function buildSkipResult(
  payload: StatePayload,
  stateFile: string,
  status: string,
  minutesSinceReply: number,
  observer: ReturnType<typeof normalizeObserverHints>,
  reason: string,
): ScanResult {
  return {
    ...buildBaseResult(payload, stateFile, status, minutesSinceReply, observer),
    should_nudge: false,
    reason,
    message: null,
  };
}

export function scan(apply: boolean, seed?: number) {
  if (!existsSync(PIPELINES)) {
    mkdirSync(PIPELINES, { recursive: true });
  }

  const files = readdirSync(PIPELINES)
    .filter((name) => name.endsWith(".json"))
    .sort();

  const results: ScanResult[] = [];

  for (const fileName of files) {
    const statePath = path.join(PIPELINES, fileName);
    const payload = JSON.parse(readFileSync(statePath, "utf8")) as StatePayload;
    const status = typeof payload.status === "string" ? payload.status : "planned";

    if (!WAITING.has(status)) continue;

    const observer = normalizeObserverHints(
      (payload.observer as Record<string, unknown> | undefined) ?? undefined,
      parseActionKind(payload.preset, "implementation"),
    );
    const minutes = inferMinutesSinceReply(payload, statePath);

    if (isDemoTestSample(payload, statePath)) {
      results.push(
        buildSkipResult(payload, statePath, status, minutes, observer, "데모/테스트/샘플 제외"),
      );
      continue;
    }

    if (!observer.track) {
      results.push(
        buildSkipResult(payload, statePath, status, minutes, observer, "observer.track 꺼짐"),
      );
      continue;
    }

    const decision = evaluateStateFile(statePath, minutes, seed, apply);
    results.push({
      ...buildBaseResult(payload, statePath, status, minutes, observer),
      ...decision,
    });
  }

  const eligible = results
    .filter((item) => item.should_nudge)
    .sort((left, right) => {
      const leftPriority = parseNumber(left.priority, 0);
      const rightPriority = parseNumber(right.priority, 0);
      if (leftPriority !== rightPriority) return rightPriority - leftPriority;
      const leftMinutes = parseNumber(left.minutes_since_reply, 0);
      const rightMinutes = parseNumber(right.minutes_since_reply, 0);
      if (leftMinutes !== rightMinutes) return rightMinutes - leftMinutes;
      return String(left.job_id ?? "").localeCompare(String(right.job_id ?? ""));
    });

  return {
    checked: results.length,
    eligible: eligible.length,
    top: eligible[0] ?? null,
    results,
  };
}

function parseArgs(argv: string[]): { apply: boolean; seed?: number } {
  const toFind = (key: string): string | undefined => {
    const idx = argv.indexOf(`--${key}`);
    if (idx === -1 || idx + 1 >= argv.length) return undefined;
    return argv[idx + 1];
  };
  const has = (key: string): boolean => argv.includes(`--${key}`);
  return {
    apply: has("apply"),
    seed: toFind("seed") ? Number(toFind("seed")) : undefined,
  };
}

function main(): void {
  const args = parseArgs(process.argv.slice(2));
  const result = scan(args.apply, args.seed);
  process.stdout.write(`${JSON.stringify(result, null, 2)}\n`);
}

if (import.meta.main) {
  main();
}
